weave-js/src/core/model/graph/serializeAllValues.ts

```diff
@@ -0,0 +1,155 @@
+// Handle serialization/deserialization of CG, for remote execution
+
+// Compared to serialize.ts, which only converts nodes to references,
+// this version replaces every single value with a reference.
+// Values which are identical (including arrays and objects)
+// are always deduplicated when serialized.
+
+// Unserialized:
+// {
+//   num: 1,
+//   str: "asdf",
+//   arr: [1, "asdf"],
+//   obj: {
+//     a: 1,
+//     b: "asdf"
+//   }
+// }
+
+// Serialized:
+// {
+//   nodes: [
+//     1,
+//     "asdf",
+//     [0, 1],
+//     {a: 0, b: 1},
+//     {
+//       num: 0,
+//       str: 1,
+//       arr: 2,
+//       obj: 3
+//     }
+//   ],
+//   targetNodes: [4]
+// }
+
+import type {EditingNode} from './editing';
+
+type SerializedRef = number;
+
+interface BatchedGraphs {
+  nodes: any[];
+  targetNodes: SerializedRef[];
+}
+
+type Serializer = {
+  nextID: SerializedRef;
+  refByHash: Map<string, SerializedRef>;
+  values: any[];
+};
+
+function createSerializer(): Serializer {
+  return {
+    nextID: 0,
+    refByHash: new Map(),
+    values: [],
+  };
+}
+
+function hash(v: any): string {
+  // When hashing an object, we must keep the keys in a consistent order.
+  // Otherwise, identical objects may be hashed into different key orders.
+  const isObject = v != null && typeof v === `object` && !Array.isArray(v);
+  if (isObject) {
+    // This does not properly stringify nested objects, but that's ok because
+    // the object values should already be replaced with serialized refs.
+    return JSON.stringify(v, Object.keys(v).sort());
+  }
```

Comment:
i would not normally expect so see a `set`, but in theory could be possible. We might want to stringify the elements, then sort the stringified elements, then stringify the set of strings.

Reply:
Are you saying that we have JS `Set`s in the graph to serialize? Interesting. How does the old `serialize` code handle that? It must transform it into something else because `JSON.stringify`ing a set always just yields `"{}"`.